### PR TITLE
v3.1

### PR DESCRIPTION
**Pull request by rafmudaf**
**Feature or improvement description**
Merge `develop` to `main` for v3.1 release.

**Additional supporting information**
Proposed release notes are given below.

## Release TODO
- Update the version in
    - [x] README.rst
    - [x] docs/index.rst
    - [x] floris/VERSION
- [x] ~Verify readthedocs builds correctly~ GH Pages is actually using the develop branch
- [ ] Create a tag in the NREL/FLORIS repository

## Release Notes

#358 Feature: add IEA 15MW turbine
#373 / #407 Add no_wake option to floris calculations
#384 Add normal vector to CutPlane to invert plot axis
#369 Redesign AEP calculation function
#385 Bug fix for yaw optimizer: revise verify_convergence to support multiple wind speeds
#382 Add cut out and default reference option
#388 Save yaw angles for reset in plane visualizations
#386 Documentation improvements
#393 Remove energy ratio tools in favor of FLASC
#308 Add UncertaintyInterface for modeling inflow wind direction uncertainty
#381 Fix for repeated calls to calculate_wake returning incorrect values
#387 Bug fix and formatting for legacy reader for FLORIS V2.4 input files
#395 / #406 Add 'none' submodel option for wake deflection, deficit and turbulence
#404 Add command line interface to the legacy input converter
#396 Adding the TurbOPark wake velocity deficit model
#408 Add GH Actions to run examples
#410 Create all Grid data at initialization

### CODE REVIEW

File: floris/simulation/wake_deflection/none.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        self,
+        grid: Grid,
+        flow_field: FlowField,
+    ) -> Dict[str, Any]:
+
+        kwargs = dict(
+            x=grid.x,
+            y=grid.y,
+            z=grid.z,
+            freestream_velocity=flow_field.u_initial,
```
Comment (bayc) [inline floris/simulation/wake_deflection/none.py:41]: These need to be updated to the `_sorted` versions for x, y, z, and u_initial

File: README.md [hunk 2]
```diff
@@ -4,22 +4,25 @@ FLORIS is a controls-focused wind farm simulation software incorporating
 steady-state engineering wake models into a performance-focused Python
 framework. It has been in active development at NREL since 2013 and the latest
 release is [FLORIS v3.0](https://github.com/NREL/floris/releases/tag/v3.0)
```
[the remark this answers is not in the document]
Comment (rafmudaf) [reply README.md]: Ah gosh I had made this change on my fork instead of here. Fixed.

[the remark this answers is not in the document]
Comment (bayc) [reply README.md]: Ah, got it! I like this pull request for a version update kind of thing. Helps to see all the changes together.

**Comment by bayc** [comment]
@rafmudaf We should probably add tests for the `none` configurations? The missing `_sorted` versions may have been caught with a test.

**Comment by bayc** [comment]
@rafmudaf I made an attempt at defining some tests for the "none" models, although I just realized I forgot about the wake_turbulence "none" model. If you can look at the tests, and this form makes sense, then we can add one for the wake_turbulence "none" model as well.

File: floris/simulation/wake_deflection/none.py (hunk 1 above)
Comment (bayc) [reply floris/simulation/wake_deflection/none.py]: Added.

File: floris/simulation/wake_velocity/none.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        self,
+        grid: Grid,
+        flow_field: FlowField,
+    ) -> Dict[str, Any]:
+
+        kwargs = dict(
+            x=grid.x,
+            y=grid.y,
+            z=grid.z,
+            u_initial=flow_field.u_initial,
```
[the remark this answers is not in the document]
Comment (bayc) [reply floris/simulation/wake_velocity/none.py]: Added.

**Comment by rafmudaf** [comment]
I think what you're intending to do in the unit test should go into a None-model regression test. It's also not a bad idea to have unit-level tests for the model functions, but that would look different in that we would try to get as close to testing only the `model.function()` as possible and we might throw lots of unphysical numbers at the model to test for robustness. A regression test here would check that the correct form of the grid variables and flow field variables have been initialized and are used correctly.

File: examples/inputs/turbopark.yaml [hunk 4]
```diff
@@ -0,0 +1,92 @@
+
+name: TurbOPark
+description: Three turbines using TurbOPark model
+floris_version: v3.0.0
+
+logging:
+  console:
+    enable: false
```
Comment (Bartdoekemeijer) [inline examples/inputs/turbopark.yaml:8]: This is `false` while for all others its `true` by default

**Comment by Bartdoekemeijer** [comment]
Thanks for opening this PR, @rafmudaf!

I'm finding that `00_getting_started.ipynb` and `07_calc_aep_from_rose.py` no longer work. All the other examples work fine, and the `UncertaintyInterface` and `none` submodels work well too. Could someone verify examples 00 and 07 on their machine?

**Comment by paulf81** [comment]
hi @Bartdoekemeijer  thanks for checking, can confirm that both of these fail on my (mac) machine too.

For the case of example 7, the issue seems to arise from the no_wake call:

```
Farm AEP (with cut_in/out specified): 57.895 GWh
Traceback (most recent call last):
  File "/Users/pfleming/Desktop/git_tools/floris/examples/07_calc_aep_from_rose.py", line 84, in <module>
    aep_no_wake = fi.get_farm_AEP(freq, no_wake=True)
  File "/Users/pfleming/Desktop/git_tools/floris/floris/tools/floris_interface.py", line 692, in get_farm_AEP
    farm_power[:, conditions_to_evaluate] = self.get_farm_power()
  File "/Users/pfleming/Desktop/git_tools/floris/floris/tools/floris_interface.py", line 607, in get_farm_power
    turbine_powers = self.get_turbine_powers()
  File "/Users/pfleming/Desktop/git_tools/floris/floris/tools/floris_interface.py", line 549, in get_turbine_powers
    turbine_type_map=self.floris.farm.turbine_type_map,
AttributeError: 'Farm' object has no attribute 'turbine_type_map'
```
Where line 84 is:

```
aep_no_wake = fi.get_farm_AEP(freq, no_wake=True)
```


In the notebook, I'm having trouble trying to display the 3D grid, maybe the structure of grid has changed?
image

**Comment by rafmudaf** [comment]
I think the problem in the notebook is that `grid.x` doesn't exist at that point since this is before doing the computation; that is, `grid.finalize` hasn't run so only `grid.x_unsorted` exists at that point

**Comment by rafmudaf** [comment]
I'm adding a GitHub Action to run the examples for each commit to check whether they run (not evaluating the results).

**Comment by Bartdoekemeijer** [comment]
I really like that idea, @rafmudaf! And thank you for rerunning those examples @paulf81. For `07_calc_aep_from_rose.py`, the code seems to be missing `self.floris.farm.turbine_type_map` and a couple other variables. I'm opening a pull request now that solves the `no_wake=True` problems.

**Comment by bayc** [comment]
> I really like that idea, @rafmudaf! And thank you for rerunning those examples @paulf81. For `07_calc_aep_from_rose.py`, the code seems to be missing `self.floris.farm.turbine_type_map` and a couple other variables. I'm opening a pull request now that solves the `no_wake=True` problems.

Beyond x, y, z, and u, v, w, there are several other variables that now have `_sorted` versions, turbine_type_map being one of those. I'll look at the code to see if there is a proper way to initialize these values for the users before any calculation is done.
